Compatibility for Optional -> AnyHashable casts

tbkka opened the pull request:
As part of making casting more consistent, the behavior of Optional -> AnyHashable casts
was changed in some cases.  This PR restores the old behavior.

Background: Most of the time, casts from String? to AnyHashable get optimized
to just injects the String? into the AnyHashable, so the following
has long been true and remains true in Swift 5.4:
```
let s = "abc"
let o: String? = s

// Next test is true because s is promoted to Optional<String>
print(s == o)

// Next test is false: Optional<String> and String are different types
print(s as AnyHashable == o as AnyHashable)
```

But when casts ended up going through the runtime, Swift 5.3 behaved
differently, as you could see by casting a dictionary with `String?` keys (in
the generic array code, key and value casts always use the runtime logic).  In
the following code, both print statements behave differently in current builds than
before:
```
let a: [String?:String] = ["Foo":"Bar"]
let b = a as [AnyHashable:Any]
print(b["Foo"] == "Bar")  // Works in Swift 5.3 and before
print(b["Foo" as String?] == "Bar") // Broken in Swift 5.3 and before
```

Swift 5.3 and before: The `String?` keys would get unwrapped to `String` before being injected into AnyHashable.  This allows the first to work but strangely breaks the second.

Behavior after casting overhaul: The `String?` keys do not get unwrapped.  This breaks the first but makes the second work.

This change restores the behavior for this specific case to the Swift 5.3 behavior.

TODO: The long-term goal, of course, is for `AnyHashable("Foo" as String?)` to test equal to `AnyHashable("Foo")` (and hash the same, of course).  See SR-9047 for details.  Once that's done, we won't need to unwrap optionals in this case and can revert this change so the runtime behavior will precisely match the behavior of compiler-optimized casts.

Resolves rdar://73301155

tbkka committed:
Compatibility for Optional -> AnyHashable casts

As part of making casting more consistent, the behavior of Optional -> AnyHashable casts
was changed in some cases.  This PR provides a hook for re-enabling the old behavior
in certain contexts.

Background: Most of the time, casts from String? to AnyHashable get optimized
to just injects the String? into the AnyHashable, so the following
has long been true and remains true in Swift 5.4:
```
let s = "abc"
let o: String? = s

// Next test is true because s is promoted to Optional<String>
print(s == o)

// Next test is false: Optional<String> and String are different types
print(s as AnyHashable == o as AnyHashable)
```

But when casts ended up going through the runtime, Swift 5.3 behaved
differently, as you could see by casting a dictionary with `String?` keys (in
the generic array code, key and value casts always use the runtime logic).  In
the following code, both print statements behave differently in Swift 5.4 than
before:
```
let a: [String?:String] = ["Foo":"Bar"]
let b = a as [AnyHashable:Any]
print(b["Foo"] == "Bar")  // Works before Swift 5.4
print(b["Foo" as String?] == "Bar") // Works in Swift 5.4 and later
```

Old behavior: The `String?` keys would get unwrapped to `String` before being injected into AnyHashable.  This allows the first to work but strangely breaks the second.

New behavior: The `String?` keys do not get unwrapped.  This breaks the first but makes the second work.

TODO: The long-term goal, of course, is for `AnyHashable("Foo" as String?)` to
test equal to `AnyHashable("Foo")` (and hash the same, of course).  In that
case, all of the tests above will succeed.

Resolves rdar://73301155

tbkka commented:
@swift-ci Please test

tbkka committed:
Fix spelling of DynamicCastResult

tbkka commented:
@swift-ci Please test

tbkka committed:
Accessing nil-keyed elements is trickier than I thought

tbkka commented:
@swift-ci Please test

lorentey commented on stdlib/public/runtime/DynamicCast.cpp:
```diff
@@ -712,6 +712,14 @@ struct ObjCBridgeMemo {
+  // AnyHashable avoids optionals before Swift 5.4
+  // TODO: This is admittedly awkward, but I need to get this changed quickly
+  // I'll come back and clean this up later.
+  return !unexpectedNullIsFatal();
```
I think this should just return true for now, without doing a linked-on-or-after check -- otherwise currently working code that bridges optional-keyed dictionaries through AnyHashable will subtly break when recompiled with the SDKs that ship with 5.4.

The issue with Optionals vs AnyHashable is tracked in https://bugs.swift.org/browse/SR-9047 (with a WIP fix in https://github.com/apple/swift/pull/20523). The fix used to be blocked by our inability to add versioned conformances. Slava resolved that in https://github.com/apple/swift/pull/34565, so I'm hoping we can land the fix in the release after 5.4.

Once SR-9047 gets fixed, we can add the link-on-or-after check to get the better casting behavior without breaking dictionary bridging.

tbkka replied:
I agree.  Done.

lorentey reviewed (approved):
LGTM! I think we should preserve the old behavior unconditionally for now though; details below.

tbkka commented:
@swift-ci Please test Windows

tbkka committed:
Don't link-check the Optional/AnyHashable workaround

This restores the earlier behavior of Optionals cast to
AnyHashable, so that [String?:Any] dictionaries cast
to [AnyHashable:Any] can be indexed by plain String
keys.

This is a little problematic because it's not consistent with the
compiler-optimized casts.

But the ability to index such dictionaries by plain String
keys seems important to preserve.  SR-9047 will expand
Optional/AnyHashable interoperability so that such
indexing works without this special case.

An earlier proposal would link-check this, changing the behavior
depending on the caller.  But it's not really workable to
change the behavior seen by intermediate frameworks depending
on the app they're being called by.

tbkka commented:
After thinking on it through the weekend, I think Karoy's right.  I've changed it to unconditionally preserve the previous behavior.  (And updated the tests to match.)

tbkka commented:
@swift-ci Please test and merge

tbkka commented:
@swift-test Please test Windows platform

tbkka commented:
@swift-ci Please test Windows platform
